### ISSUE TITLE
OpenSCAD Benchmark Suite proposal and discussion

### ISSUE THREAD

**Issue by thehans** [openscad/openscad#3931]
It has been brought up in various discussions that we could use a good performance benchmark suite.
This issue is for discussing what these benchmarks should cover, and to provide a place to collect some scripts to get us started.

### Overall Goals
The main motivations for this benchmark suite would be:
* Help identify and analyze any current inefficiencies in OpenSCAD (possibly platform-specific even, eg. #2744)
* Regularly check for unintended performance regressions between versions.
* Simplify testing and demonstration of newly developed performance optimizations (e.g. #3641, VBO preview, future multithreading possibilities, etc).

For simplicity and ease of contribution I think the suite should mainly just be a collection of scad files (whole-app "integration testing" as opposed to micro benchmarks).
Some of my thoughts general guildelines:
* Most of these scad benchmarks should be a minimal, self contained file (with exceptions for one or two "large project" tests which stress heavy `include` and/or `use`.  also see related #3884)
* Target runtime of individual tests should somewhere around the range of 1 to 10 seconds.
* In order to support a wide range of hardware, each script should accept a `complexity` parameter, via `-D ` command line argument, which should attempt to scale the run time as close to linearly as possible;  be that by altering `$fn`, a number of loop iterations, geometry count, etc., whatever is appropriate for the specific test.
    This would be a convenience so that slower hardware does not drag on too long, or much faster hardware has long enough runtimes for comparisons that are not "in the noise".

### Test Coverage Goals
As far as the collection of tests, I think ideally (eventually) this would include, but not necessarily limited to, one for every **builtin** module and function.
But for a start, and probably 80-90% of the usefulness, just having a few good boolean ops tests would be great.

- [ ] F6 Render of Boolean CSG Ops: `union`, `difference`, `intersection`
    For each of these primary boolean operations, I think it would be good to cover a range of different forms; classifying problems based on a combination of geomtry complexity (# of facets) and number of geometries.
    (See **CSG Classification** section below for details)
- [ ] F6 Render of other operations: `minkowski`, `hull`, `resize`, `projection`, etc.
- [ ] Core language (expression evaluation):
     - heavy function calls: builtin functions, in addition to user-defined (using recursion; TCO and non-TCO)
     - Higher-order function literals
     - Building large vectors (also strings?) via list-comprehensions, and recursive concatenations, etc.
     - Some user-implemented algorithms (my https://github.com/thehans/funcutils library could be a good source for some algorithms, or as a base to build more complex algos upon)
- [ ] F5 Preview (OpenCSG) performance tests.  Create some worst-case tests, eg inspired by [csg term normalization issues](https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/FAQ#Why_is_the_preview_so_slow?)?
- [ ] echo / warning performance. This may require some way to perform [GUI automation](https://github.com/openscad/openscad/issues/2829) for meaningful benchmarks.
- [ ] One or two "large" projects with many `include`s and/or `use`s (also see https://github.com/openscad/openscad/issues/3884 )
- [ ] A few practical / real-world model tests (as opposed to mostly "synthetic" scripts explicitly for stressing parts of OpenSCAD)
- [ ] Animation frame dumps?

### Classification of CSG Problems
The idea behind this is that depending on the input geometries, testing boolean CSG ops could have wildly different performance profiles (i.e. a particular optimization might help greatly in one case and hardly at all in another).
For the following examples, the "size" of geometries (Small / Medium / Large) refers to the relative facet count and not scale of dimensions.  e.g. Small could mean a `square` or `cube`, and Large geometry could be a `sphere` with high `$fn`
* Operations between 2 Large geometries
* Operations between 1 Large geometry and Many Small ones
* Operations between Very Many Small ones, or Many Medium sized ones.
* Geometries which are disjoint vs actually intersecting.
* Others ?

### Bench Runner
We will need a bench runner script, likely in python for cross platform capability.  Able to support the following:
- [ ] Ability to run on all supported platforms: Windows / Linux / MacOS
- [ ] Collect and report data from individual scad
  - [ ] Raw CSV output with: test name, run time (),
    - [ ] Run time per case
    - [ ] Peak memory usage per case. This may be difficult to do fully cross platform, but could be very useful diagnostics.
- [ ] Script command line options:
  - [ ] (OpenSCAD) executable path, so that various versions can be tested and compared.
  - [ ] experimental features to enable across the run.
  - [ ] Complexity scale
  - [ ] Existing report to compare against
  - [ ] Parallel jobs?  Not sure if this should be supported, as it could skew results randomly.
- [ ] Verify output correctness? much like main test suite. (while not the main purpose of this suite, could act as a supplementary test suite)

### Implementation / Organization of Suite
The benchmarks *could* potentially be added to a subdirectory of the main repo just as the existing test suite does,
but I think it would better to instead give the benchmark suite its own repo (owned by OpenSCAD org to make it official).
That way its not necessarily tied to a specific release or build version of the main app.

The scad files should mostly reside within the benchmark repo, with possibly some of the larger projects, or libraries being included as git submodules.

**Comment by ChrisCoxArt** [comment]
I would place GUI drawing and echo/warning at a lower priority.  GUI drawing is not easy to benchmark correctly on modern platforms (drawing command issued != visible on screen due to GPU pipelines and display refresh rates).  There are many UI drawing issues that probably need some performance monitoring, and validation with stopwatches.

The highest priority should probably be basic geometry operations that are already known to take significant time.
(in other words: focus on known problems and what people complain about first)

In the report, include as much info as possible about the system being run on, plus date, time, and configuration options (for the app and script).

Run serially, don't try to tackle running benchmarks in parallel ("that way lies madness").

Verifying correctness would be nice to have, as a smoke test - but probably not a high priority.  Maybe we can implement a simple version (check #faces and #objects) and build on it later.

**Comment by nophead** [comment]
My scrips currently measure the run time of OpenScad whenever it is calledit from the command line and my framework shows how it has changed since the last run and presents it like this:
![image](https://user-images.githubusercontent.com/566149/136523356-7cc1a7ae-8310-4183-a3b7-ffaac22e25f6.png)

I use it to compare code optimisations in my scad code, etc. I can also compare OpenSCAD versions. The big problem is the noise in the times makes it difficult to know if a change is actually faster or slower because it is real time, not CPU time. I have looked into making it CPU time but I think the only way to do that is to get OpenSCAD to output the elapsed CPU time as it finishes. I don't think the calling process can get at the total elapsed CPU time of its child.

**Comment by Scopeuk** [comment]
My understanding is "real cpu time" can be accesed through unix calls https://linux.die.net/man/2/times typically this is exposed via the "time" call on a linux cli "time openscad ..." this provides clock time as "real" but also user and system times which when summed give an "on cpu" time value. I've no doubts there are huge limitations to this method but it might be worth a shot as the effort is reasonably low

**Comment by ChrisCoxArt** [comment]
Any progress?  Anything we can help with?

**Comment by thehans** [comment]
> Any progress? Anything we can help with?

In testing out the alternative allocator "mimalloc" (PR #3943 and Issue #3930), I found that it is able to provide `time`-like functionality of printing real and system time upon termination, *in addition to* the very convenient peak and total memory allocation, by setting environment variable `MIMALLOC_SHOW_STATS=1` See the usage with an example output [from their docs here](https://github.com/microsoft/mimalloc#using-the-library).

This looks like the perfect choice to me, and much nicer than trying to poll from parent Python process as I previously suggested.  But I ran into problems getting a build which worked across all platforms, and have not been able to finalize that PR yet.  One method of linking and implementing it caused every test on mac to crash and fail, and another method I tried caused every test on windows to crash.  I was hoping for some more help testing on the crashing issues on mac (you know, that time where you ignored my "@ChrisCoxArt" for 2months or so 😛) , as I have no access to mac hardware (nor do I ever wish for any).  I can try debugging the Windows builds, but reproducing the same environment as the build server is a bit of a chore, so I've stalled out a bit.

Also I tried to get CodeQL and LGTM CI tests to ignore mimalloc as a git submodule, but they still complained after I added what I thought was the correct config to ignore it.  Would love any input on that.

Another small issue is that I'd like to have a new repo created under github.com/openscad to place these benchmarks in (hint: @t-paul 😉).  I sort of mentioned this in "Implementation / Organization of Suite" section of my original post, but maybe was too subtle.  Until then it would be fine to post some example scripts as comments in this discussion.  
I already have a small collection of scad files for benchmarking, just haven't gotten around to cleaning them up and posting them yet.

**Comment by t-paul** [comment]
There is also https://github.com/openscad/statistics-scripts but a new repo probably makes sense.

**Comment by ChrisCoxArt** [comment]
Hey, I didn't ignore it, I just never received the notification (and I even check my spam filters weekly!).
Mac hardware:  it's like linux, with actual software, standards, and support.  :-P
I'll see if I can get the mimalloc branch to build, in my copious spare time.

**Comment by pca006132** [comment]
Hi, I wonder if there is any benchmark suit that you guys use to catch regression.

It seems that the current unit tests (running ctest) are not very comprehensive. I currently need to use the benchmarks for manifold integration (https://gist.github.com/ochafik/70a6b15e982b7ccd5a79ff9afd99dbcf) for catching regression (both performance and correctness), but the set may still be a bit too small to cover all cases.

Btw, I wonder if there is a way to speed up openscad compilation or reduce the memory usage? On my machine I can only use 10 concurrent jobs for compilation, otherwise it will use up my 32GB of RAM and start using the swap which slows everything down. I think CMake supports PCH and unity build, not sure if you guys tried it before.

It is known that manifold does not support unity build because we use anonymous namespace functions everywhere, where unity build will cause name collision. And it is not worth it for our case to modify them and support unity build, because manifold compiles pretty quickly and does not have huge header dependencies. For OpenSCAD, it may help to enable unity build as boost and CGAL are fairly large...
